### PR TITLE
remove multi-threaded downloading

### PR DESCRIPTION
**Pull request by AndreyNikiforov**
feat: removed multi-threaded downloading and added deprecation notice to --threads-num parameter #180, #188

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -118,7 +118,7 @@ pip install icloudpd
                                         prints log messages on separate lines
                                         (Progress bar is disabled by default if
                                         there is no tty attached)
-        --threads-num INTEGER RANGE     Number of cpu threads (default: 1)
+        --threads-num INTEGER RANGE     Number of cpu threads -- depricated. To be removed in future version
```
Comment (ndbroadbent) [inline README.md:121]: Minor typo here - should be "deprecated"

File: icloudpd/base.py [hunk 2]
```diff
@@ -194,7 +188,7 @@
               )
 @click.option(
     "--threads-num",
-    help="Number of cpu threads (default: 1)",
+    help="Number of cpu threads -- depricated. To be removed in future version",
```
Comment (ndbroadbent) [inline icloudpd/base.py:191]: also here "deprecated"

**Comment by ndbroadbent** [comment]
Sure, I just saw this follow-up comment about the download times: https://github.com/icloud-photos-downloader/icloud_photos_downloader/pull/171#issuecomment-715300917

So that sounds fine to me!

File: tests/test_download_photos.py [hunk 3]
```diff
@@ -322,7 +322,6 @@ def test_until_found(self):
         files_to_skip.append(("2018/07/30/IMG_7400.JPG", "photo", 2308885))
         files_to_skip.append(("2018/07/30/IMG_7400-medium.MOV", "photo", 1238639))
         files_to_skip.append(("2018/07/30/IMG_7399.JPG", "photo", 2251047))
-        files_to_download.append(("2018/07/30/IMG_7399-medium.MOV", "photo"))   # TODO this should not be downloaded, correct?
```
Comment (menkej) [inline tests/test_download_photos.py:325]: Why did you remove this? Is that the file that causes the test issues for you?

**Review by menkej — COMMENTED** [review]
Fine with me - apart from the typo, but I'm currently running some tests on multithreading, so let's wait some days before we really remove it.

File: tests/test_download_photos.py (hunk 3 above)
Comment (AndreyNikiforov) [reply tests/test_download_photos.py:325]: That file should never be downloaded, because all processing should stop after three skips. i added check to the test.

File: README.md (hunk 1 above)
Comment (AndreyNikiforov) [reply README.md]: fixed

**Comment by AndreyNikiforov** [comment]
> Fine with me - apart from the typo, but I'm currently running some tests on multithreading, so let's wait some days before we really remove it.

Current (1.6.2) multi-threaded downloading implementation has a number of shortcomings that cause a majority of issue reported for the project since I started paying more attention to it:
- deduplication does not work with threads > 1, we changed default to 1 to reduce chances of ppl hitting this issue
- until_found is inconsistent (btw it forces threads == 1), causing failing tests (==false alarms); this is because current pub-sub implementation checks counter at publisher and moving this check to subscriber requires additional back pressure implementation

We did confirm in a number of tests that multi threaded downloading does not improve speed on fast and slow connections. Anybody should be able to repeat tests and validate for themselves using 1.6.2. Here are things to keep in mind while testing multi-threaded vs single-threaded:
- iCloud account should not have any duplicates
- test should not use --unit-found parameter

If we are to fix/re-implement parallel downloading, we'll have to keep the following in mind:
- deduplication should take both in-flight tasks and files on disk into account
- until-found may be really tricky to implement to avoid any unnecessary processing/downloading
- it would be most likely more resource efficient to use event loop-based IO concurrency like asyncio

With all that in mind, I feel it would be better to release this feature as 1.7.0, continue any multi-threaded downloading testing using 1.6.2, and revisit parallel downloading if tests show any promises and we have resources to implement parallel downloading correctly. @menkej will that work for you?

**Comment by jmonster** [comment]
@AndreyNikiforov was this feature re-added? The cli arg `--help` still advertises this feature (as does the README). I'm not asking you to put it back, but if it is _is_ supposed to work, it doesn't seems broken

https://github.com/icloud-photos-downloader/icloud_photos_downloader/search?q=threads_num

**Comment by AndreyNikiforov** [comment]
> was this feature re-added? 

What makes you think so?

> The cli arg `--help` still advertises this feature (as does the README). I'm not asking you to put it back, but if it is _is_ supposed to work, it doesn't seems broken

What do you mean by `advertising`? Help in 1.7.2 says that feature is deprecated and will be removed. It is not broken, it is just noop (==does nothing).
